### PR TITLE
RISC-V: Add kernel stack

### PR DESCRIPTION
**Pull request by pussuw**
## Summary
Adds support for a separate kernel stack, needed by CONFIG_BUILD_KERNEL=y
## Impact
No impact to existing code
## Testing
Tested with icicle:knsh (CONFIG_BUILD_KERNEL=y) / not released yet

This includes more missing pieces to get CONFIG_BUILD_KERNEL=y for RISC-V. Code is basically copy&pasted from the ARMv7 implementation.

### CODE REVIEW

File: arch/risc-v/src/common/addrenv.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#ifdef CONFIG_ARCH_ADDRENV
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Aligned size of the kernel stack */
+
+#ifdef CONFIG_ARCH_KERNEL_STACK
+#  define ARCH_KERNEL_STACKSIZE ((CONFIG_ARCH_KERNEL_STACKSIZE + 7) & ~7)
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/addrenv.h:42]: could we reuse STACK_ALIGNMENT and STACK_ALIGN_XXX macro in riscv_internal.h?

File: arch/risc-v/src/common/riscv_swint.c [hunk 2]
```diff
@@ -440,6 +472,22 @@ int riscv_swint(int irq, void *context, void *arg)
...
+#ifdef CONFIG_ARCH_KERNEL_STACK
+          /* We must enter here be using the user stack.  We need to switch
+           * to back to the kernel user stack before returning to the kernel
+           * mode signal trampoline.
+           */
+
+          if (rtcb->xcp.kstack != NULL)
+            {
+              DEBUGASSERT(rtcb->xcp.kstkptr != NULL &&
+                          (uintptr_t)rtcb->xcp.ustkptr == regs[REG_SP]);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_swint.c:485]: is it true?

File: arch/risc-v/src/common/riscv_addrenv_kstack.c [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+ *   tcb - The TCB of the thread that requires the kernel stack.
+ *
+ * Returned Value:
+ *   Zero (OK) on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+
+int up_addrenv_kstackalloc(FAR struct tcb_s *tcb)
+{
+  DEBUGASSERT(tcb && tcb->xcp.kstack == 0);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_addrenv_kstack.c:63]: 0->NULL

File: arch/risc-v/src/common/riscv_addrenv_kstack.c [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+  /* Allocate the kernel stack */
+
+  tcb->xcp.kstack = (uintptr_t *)kmm_memalign(8, ARCH_KERNEL_STACKSIZE);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_addrenv_kstack.c:67]: should we manually algin here like up_create_stack/up_use_stack?
8->STACK_ALIGNMENT

File: arch/risc-v/include/irq.h [hunk 5]
```diff
@@ -517,6 +517,32 @@ struct xcptcontext
...
+
+#ifdef CONFIG_ARCH_ADDRENV
+#ifdef CONFIG_ARCH_STACK_DYNAMIC
+  /* This array holds the physical address of the level 2 page table used
+   * to map the thread's stack memory.  This array will be initially of
+   * zeroed and would be back-up up with pages during page fault exception
+   * handling to support dynamically sized stacks for each thread.
+   */
+
+  uintptr_t *ustack[ARCH_STACK_NSECTS];
```
[the remark this answers is not in the document]
Comment (pussuw) [reply arch/risc-v/include/irq.h:530]: I can remove this yes. It was just copypasted from arch/arm/include/armv7-a/irq.h

File: arch/risc-v/src/common/addrenv.h (hunk 1 above)
Comment (pussuw) [reply arch/risc-v/src/common/addrenv.h:42]: Yes, good point. Will fix this. The alignment is also wrong, as 16B alignment is required.

File: arch/risc-v/src/common/riscv_addrenv_kstack.c (hunk 3 above)
Comment (pussuw) [reply arch/risc-v/src/common/riscv_addrenv_kstack.c:63]: Sure!

File: arch/risc-v/src/common/riscv_addrenv_kstack.c (hunk 4 above)
Comment (pussuw) [reply arch/risc-v/src/common/riscv_addrenv_kstack.c:67]: Yes I'll change 8->STACK_ALIGNMENT. The 8B alignment is wrong anyway.

File: arch/risc-v/src/common/riscv_swint.c (hunk 2 above)
Comment (pussuw) [reply arch/risc-v/src/common/riscv_swint.c:485]: This is also copied from arch/arm/src/armv7-a/arm_syscall.c

I have not yet had time to test the userspace signal handler, so cannot verify if this works. To me it looks fine.

The comment at least is true, this syscall must come from userspace and the user stack must be in use.

Comment (xiaoxiang781216) [reply arch/risc-v/src/common/riscv_swint.c:485]: This first condition(rtcb->xcp.kstkptr != NULL) looks good, but the second((uintptr_t)rtcb->xcp.ustkptr == regs[REG_SP]) look strange.

File: arch/risc-v/src/common/riscv_swint.c [hunk 6]
```diff
@@ -415,6 +429,24 @@ int riscv_swint(int irq, void *context, void *arg)
...
+           * information to make this decision.
+           */
+
+          if (rtcb->xcp.kstack != NULL)
+            {
+              DEBUGASSERT(rtcb->xcp.kstkptr == NULL &&
+                          rtcb->xcp.ustkptr != NULL);
+
+              rtcb->xcp.kstkptr = (uintptr_t *)regs[REG_SP];
+              regs[REG_SP]      = (uintptr_t)rtcb->xcp.ustkptr;
```
Comment (pussuw) [inline arch/risc-v/src/common/riscv_swint.c:447]: I think the assert test you mentioned below is to ensure that the stack pointer has been fully unwound to what was set here?

File: arch/risc-v/src/common/riscv_swint.c (hunk 2 above)
Comment (pussuw) [reply arch/risc-v/src/common/riscv_swint.c:485]: I assume the test is here to ensure that the user stack has been fully unwound by the userspace signal handler, i.e. when we return here the SP when dispatching the signal handler should be the same as when returning from the signal handler.

I cannot say for sure as I did not write this code.

Comment (xiaoxiang781216) [reply arch/risc-v/src/common/riscv_swint.c:485]: Ok, let keep until the assert trigger

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**